src/box/sql/show.c

```diff
@@ -0,0 +1,415 @@
+static void
+sql_desc_initialize(struct sql_desc *desc)
+{
+	desc->size = BUF_INITIAL_SIZE;
+	desc->len = 0;
+	desc->buf = sql_xmalloc0(desc->size);
+	desc->err_size = BUF_INITIAL_SIZE;
+	desc->err_len = 0;
+	desc->err_buf = sql_xmalloc0(desc->err_size);
+}
+
+/** Append a new string to the object used to accumulate a statement. */
+static void
+sql_desc_append(struct sql_desc *desc, const char *fmt, ...)
+{
+	char *buf = &desc->buf[desc->len];
+	size_t size = desc->size - desc->len;
+	va_list ap;
+	va_start(ap, fmt);
+	int rc = vsnprintf(buf, size, fmt, ap);
+	va_end(ap);
+	assert(rc >= 0);
+	if ((size_t)rc < size)
+		goto end;
+
+	while ((size_t)rc + desc->len >= desc->size)
+		desc->size *= 2;
+	desc->buf = sql_xrealloc(desc->buf, desc->size);
+	buf = &desc->buf[desc->len];
+	size = desc->size - desc->len;
+	va_start(ap, fmt);
+	rc = vsnprintf(buf, size, fmt, ap);
+	va_end(ap);
+end:
+	desc->len += rc;
+	assert(rc >= 0 && desc->len < desc->size);
+}
+
+/** Append a new error to the object used to accumulate a statement. */
+static void
+sql_err_append(struct sql_desc *desc, const char *type, const char *name)
+{
+	const char *fmt;
+	if (desc->err_len == 0)
+		fmt = "/* %s \"%s\" cannot be defined in this statement. */";
+	else
+		fmt = "\n/* %s \"%s\" cannot be defined in this statement. */";
+	char *buf = &desc->err_buf[desc->err_len];
+	size_t size = desc->err_size - desc->err_len;
+	int rc = snprintf(buf, size, fmt, type, name);
+	assert(rc >= 0);
+	if ((size_t)rc < size)
+		goto end;
+
+	while ((size_t)rc + desc->err_len >= desc->err_size)
+		desc->err_size *= 2;
+	desc->err_buf = sql_xrealloc(desc->err_buf, desc->err_size);
+	buf = &desc->err_buf[desc->err_len];
+	size = desc->err_size - desc->err_len;
+	rc = snprintf(buf, size, fmt, type, name);
+end:
+	desc->err_len += rc;
+	assert(rc >= 0 && desc->err_len < desc->err_size);
+}
+
+/** Finalize a described statement. */
+static char *
+sql_desc_finalize(struct sql_desc *desc)
+{
+	if (desc->err_len > 0) {
+		if (desc->len > 0)
+			sql_desc_append(desc, "\n");
+		sql_desc_append(desc, desc->err_buf);
+	}
+	char *buf = desc->buf;
+	sql_xfree(desc->err_buf);
+	memset(desc, 0, sizeof(*desc));
+	return buf;
+}
+
+/** Add a field foreign key constraint to the statement description. */
+static void
+sql_describe_field_foreign_key(struct sql_desc *desc,
+			       const struct tuple_constraint_def *cdef)
+{
+	assert(cdef->type == CONSTR_FKEY && cdef->fkey.field_mapping_size == 0);
+	const struct space *foreign_space = space_by_id(cdef->fkey.space_id);
+	assert(foreign_space != NULL);
+	const struct tuple_constraint_field_id *field = &cdef->fkey.field;
+	if (field->name_len > 0) {
+		uint32_t id;
+		uint32_t hash = field_name_hash(field->name, field->name_len);
+		if (tuple_fieldno_by_name(foreign_space->def->dict, field->name,
+					  field->name_len, hash, &id) != 0) {
+			sql_err_append(desc, "Foreign key", cdef->name);
+			return;
+		}
```

Comment:
This case is checked in the test on line 154. This is when the outer space does not have a field referenced by the local space. This situation is possible in Lua, because the core FK checks the availability of a field at run time, not at creation.

Reply:
But you are not **doing** the create-table here, are you? You just **generate** the statement. You don't execute it. So what is the problem anyway? You have the name - you can use it. You check it now in 3 places (which I would recommend to extract into a function if you intend to keep them even after this comment), but never use the resulting fieldno. Because the field name was enough. Why care in this `SHOW` statement if the name matches with any fieldno if you don't use it? If this field will be missing at the time of the resulting create-table execution (if it will be executed at all), then just let it fail there.

I can expand my reasoning - *"why try to guess what CREATE TABLE fails at, if you are not doing it here?"*. You only need to serialize the definition precisely. Fail only when it is not possible (like field type mismatch in index and space). This concerns not only FK fieldnos. There might be other restrictions which CREATE TABLE has and you do not account them here. For instance, we have a max column count limit - `SQL_MAX_COLUMN`. You don't check it here, but the resulting `box.execute(res.rows[1][1])` will fail if you serialized a non-SQL space with more columns.

Up to you. Those checks just seem as harmless as useless.

Reply:
Removed this check.